internal/nix/nix.go

```diff
@@ -84,31 +84,72 @@ func parseInfo(pkg string, data []byte) *Info {
 	return nil
 }
 
-type varsAndFuncs struct {
-	Functions map[string]string   // the key is the name, the value is the body.
+type printDevEnvOut struct {
 	Variables map[string]variable // the key is the name.
 }
+
 type variable struct {
 	Type  string // valid types are var, exported, and array.
 	Value any    // can be a string or an array of strings (iff type is array).
 }
 
+type PrintDevEnvArgs struct {
+	FlakesFilePath       string
+	PrintDevEnvCachePath string
+	UsePrintDevEnvCache  bool
+}
+
 // PrintDevEnv calls `nix print-dev-env -f <path>` and returns its output. The output contains
 // all the environment variables and bash functions required to create a nix shell.
-func PrintDevEnv(ctx context.Context, nixShellFilePath, nixFlakesFilePath string) (*varsAndFuncs, error) {
+func PrintDevEnv(ctx context.Context, args *PrintDevEnvArgs) (*printDevEnvOut, error) {
 	defer trace.StartRegion(ctx, "nixPrintDevEnv").End()
 
-	cmd := exec.CommandContext(ctx, "nix", "print-dev-env", nixFlakesFilePath)
-	cmd.Args = append(cmd.Args, ExperimentalFlags()...)
-	cmd.Args = append(cmd.Args, "--json")
-	debug.Log("Running print-dev-env cmd: %s\n", cmd)
-	out, err := cmd.Output()
+	var data []byte
+	var err error
+	var out printDevEnvOut
+
+	if args.UsePrintDevEnvCache {
+		if _, err := os.Stat(args.PrintDevEnvCachePath); err == nil {
+			data, err = os.ReadFile(args.PrintDevEnvCachePath)
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+			if err := json.Unmarshal(data, &out); err != nil {
+				return nil, errors.WithStack(err)
+			}
+		}
```

Comment:
You can remove the `os.Stat` call and just trying reading:

```suggestion
		cachedData, err := os.ReadFile(args.PrintDevEnvCachePath)
		if err == nil {
			if err := json.Unmarshal(data, &out); err != nil {
				return nil, errors.WithStack(err)
			}
			cachedData = data
		} else if !errors.Is(err, fs.ErrNotExist) {
			return nil, errors.WithStack(err)
		}
```